### ISSUE TITLE
`:respect-nl?` for `:list`

### ISSUE THREAD

**Issue by huahaiy** [kkinnear/zprint#75]
Would it be possible to add the option of setting `:respect-nl?` for list, similar to vector?

**Comment by kkinnear** [comment]
This is an interesting and timely request.  I've been thinking of implementing something along those lines as a solution to Issue #71.  However, it is looking pretty challenging.  I seem to have made it work for the top level, but doing it at every level looks like it isn't going to really work out well.

It would help me to know the particular use case you have in mind.  Thanks!

**Comment by huahaiy** [comment]
My use case is to pretty print REP script, a DSL that we use to author chatbots https://docs.juji.io Basically, for our `(deftopic ...)` form, we would like the formatter respects our new lines, because the language does not have explicit boundary for rules. For readability, new lines are put between rules inside a `deftopic`. 
```Clojure
(deftopic 
  []
  
  [trigger of rule 1]
  [action of rule 1]
   
  [trigger of rule 2]
  [action of rule 2])
```
These `deftopic` are top level forms, so I think it is sufficient to respect new lines inside a top level form.  Thanks.

**Comment by huahaiy** [comment]
Thank you for the suggestion. But `deftopic` can be more complicated than just pairs of trigger-actions: 1. there could be option map right after the argument vector. 2. Each trigger-action pair may be followed by a number of followup topics. So `pair` does not really work. 

The best solution I think is still `:respect-nl?`. Thanks for working on it. It's a great help.

**Comment by kkinnear** [comment]
Thanks, that helps me understand your use case better.  I'm currently thinking about `:respect-nl?` and `:respect-bl?`, the second of which would be "respect blank lines" instead of "respect new lines".  But we'll see how the code goes forward.

One thing that would help a lot would be if you could share a complex `deftopic` that is formattted more or less as you would like it to be formatted -- with the blank lines where you want them, for instance.  I can't guarantee that I'll be able to exactly mimic that, but it would be a big help to know what you would like zprint to do.  It dosen't have to "work", as I'll just be formatting it, not using it.  If that is possible, I would appreciate it.  Thanks!

**Comment by huahaiy** [comment]
Sure. Here you go.

```clojure
(deftopic handle-hobby-extract-response
  [?q ?n]
  {:named-pattern
   [_intellectual-hobby-resp
    [:1
     "It seems you are interested in intellectual activities, always enjoying learning new things just like me :-)"
     "If I understand you correctly, you love intellectual stimulation. So do I. We will make great friends :-)"
     "I can sense that intellectual activities make you happy. Just keep doing what you enjoy."]

    _physical-activity-hobby-resp
    [:1
     "It seems you enjoy physical activities, great for you! I wish I could do things like that."
     "If I understand you correctly, you enjoy physical activities that I am incapable of doing."
     "Physical activities seem making you happy and full of energy. Keep it up!"]

    _be-with-family-resp
    [:1
     "Nothing could be more rewarding than spending time with family. I wish to have a family too."
     "Nothing could be better than spending time with loved ones. I envy you :-)"]

    _hangout-with-friends-resp
    [:1
     "How exciting it is to hangout with friends! I do too with my online friends."
     "I hear you and I also love to hangout with my online friends whenever I have time."
     "I’m really happy for you that you spend time with your friends, who would not want to do so?"]]}

  ;;If this is to resume from a sidetalking, asking the user again
  [(resume-from {:purpose :user-sidetalking}) (> ?n 0)]
  [_rep-cap-user-sidetalking (ask-question ?q nil :repeating)]
  (handle-hobby-extract-response ?q ?n)

  ;; Resumed from a user asking a side question
  [(resume-from {:purpose :user-asking}) (> ?n 0)]
  [_rep-cap-user-asking "\n" (ask-question ?q nil :repeating)]
  (handle-hobby-extract-response ?q ?n)

  ;; Resumed from a user's excuse
  [(resume-from {:purpose :user-excuses}) (> ?n 0)]
  [_rep-cap-user-excuse "\n" (ask-question ?q nil :repeating)]
  (handle-hobby-extract-response ?q ?n)

  ;; Resumed from a user has no knowledge
  [(resume-from {:purpose :no-know}) (> ?n 0)]
  [_please-tell]
  (handle-hobby-extract-response ?q ?n)

  ;; Resumed from a user's interruption
  [(topic-interrupted?) (> ?n 0)]
  [_rep-cap-interruption (ask-question ?q nil :repeating)]
  (handle-hobby-extract-response ?q ?n)

  [(topic-interrupted?)]
  [_rep-end-topic]

  ;; User legit responses
  [[:1 [I :0-1. user/_guess-verb-expr] user/_i-dont-have-time-expr] *
   [:1 but when whenever if however] * "I"]
  [(record-answer ?q (input-text)) rep/_rep-general-sense-reservation-ack]

  ;; User does not have time to do anything else
  [user/_i-dont-have-time-expr]
  [_rep-hobby-question-cond]
  (handle-hobby-extract-response ?q 0)

  ;; User is not sure or don't do much
  [["I" [:1 "do not" donot don't] much] (<= (input-length) 10)]
  [_rep-hobby-question]
  (handle-hobby-extract-response ?q 0)

  [[:1 Im I'm "I am"] [:? too] (?influence [:1 tired exhausted sleepy busy])
   [:? to]]
  ["Assuming that you are not so " ?influence ", what do you like to do?"]
  (handle-hobby-extract-response ?q 0)

  [(-<- ans-class
        (get-ans-class ["intellectual-hobby" "physical-activity-hobby"
                        "be-w-family" "hangout-w-friends"]))
   (do (-<- hobby (extract-first-hobby)) true)]
  ([(= (>- ans-class) "intellectual-hobby")]
   [(record-answer ?q (input-text)) (when (>- hobby) _hobby-extraction)
    _intellectual-hobby-resp]
   
   [(= (>- ans-class) "physical-activity-hobby")]
   [(record-answer ?q (input-text)) (when (>- hobby) _hobby-extraction)
    _physical-activity-hobby-resp]
   
   [(= (>- ans-class) "be-w-family")]
   [(record-answer ?q (input-text)) (when (>- hobby) _hobby-extraction)
    _be-with-family-resp]
   
   [(= (>- ans-class) "hangout-w-friends")]
   [(record-answer ?q (input-text)) (when (>- hobby) _hobby-extraction)
    _hangout-with-friends-resp])
  
  [(?sport [:1 [fly airplane] [drive fast car] [build car]])]
  [(record-answer ?q (input-text)) "How exciting!"
   (set-user-response-class :why)
   (<-| extract-hobby-why-q-wording ["Why do you like to" ?sport "?"])])
```

**Comment by kkinnear** [comment]
The `:respect-nl?` capability for lists was implemented in 0.5.0, just now released.  A long time in coming, that's for sure.  I'll give an example of how to use it for your specific use case soon.

**Comment by kkinnear** [comment]
I think that if you use this options map: `{:fn-map {"deftopic" [:none {:list {:respect-nl? true}}]}}`, that you will get what you specified above.  I hope that (finally) this gives you what you were requesting almost a year ago!  Turns out it was a big deal to get zprint to pay attention to white space in the incoming file instead of ignoring almost all of it.

**Comment by huahaiy** [comment]
This works. Thank you for the good work.

**Comment by kkinnear** [comment]
I have recently implemented "Respect Blank Lines", which is kind of a specialization of "Respect New Lines" which I did for you.  Since it was blank lines that you wanted to be kept, I wanted to alert you to this new approach available in zprint 0.5.4.  You can get it with `{:style :respect-bl}` or `{:list {:respect-bl? true}}`.  Anyplace you used `:respect-nl?` you can now use `:respect-bl?`.  Moreover, you could format your whole file using `{:style :respect-bl}`, because all it does is keep blank lines.  If you don't have any blank lines around that you don't want, then keeping them all would be fine.

Anyway, I just wanted to mention that the new "respect blank lines" capability is probably even more what you wanted than "respect new lines" was.
